tests/testsuite/metadata.rs

```diff
@@ -627,6 +627,326 @@ fn cargo_metadata_with_deps_and_version() {
         .run();
 }
 
+#[cargo_test]
+fn cargo_metadata_public_private_dependencies() {
```

Comment:
Would you be up for splitting your commit into
- Add test (passes on this commit)
- Make behavior change (will require updating the test so it still passes)

This both makes it easier to review and easier to demonstrate to others what the new feature does.  You just look at the feature commit and see how the tests have changed.

Reply:
I like this :+1:.
I've split everything into multiple commits